README
This is the README file for the lwlogger package.

This software comes with ABSOLUTELY NO WARRANTY.  See the file 'COPYING' for
details.

See the file 'INSTALL' for instructions on building lwlogger from source.

Lwlogger is a utility for logging GPIO events on a LiveWire AoIP network.



NEWS
The NEWS file for the lwlogger package.

-------------------------------------------------------------------------------
v0.7.0 -- 03/06/2017

Changes:
Timestamp Precision. Added a tenth-second field to the log timestamp
                     when logging GPIO events.

Scripting Support.  Added support for executing a script in response to
                    GPIO events.

Watchdogs.  Added support for setting a watchdog timer on each GPIO event,
            with the ability to write a log entry or run a script when such
            timer expires.

-------------------------------------------------------------------------------


src/config.h
// config.h
//
// Configuration for lwlogger(8)
//
//   (C) Copyright 2016 Fred Gleason <fredg@paravelsystems.com>
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as
//   published by the Free Software Foundation; either version 2 of
//   the License, or (at your option) any later version.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public
//   License along with this program; if not, write to the Free Software
//   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

#ifndef CONFIG_H
#define CONFIG_H

#include <map>

#include <QObject>
#include <QTimer>
#include <QString>

#include <sy/sygpio_server.h>

#define CONFIG_FILE "/etc/lwlogger.conf"

class ConfigEntry
{
 public:
  enum Role {Off=0,On=1,Pulse=2,WatchdogSet=3,WatchdogReset=4,LastRole=5};
  ConfigEntry(int srcnum);
  int sourceNumber() const;
  QString logDir(SyGpioEvent::Type type,int line) const;
  void setLogDir(SyGpioEvent::Type type,int line,const QString &str);
  QString gpioString(SyGpioEvent::Type type,Role role,int line);
  void setGpioString(SyGpioEvent::Type type,Role role,int line,
		     const QString &str);
  QString gpioAction(SyGpioEvent::Type type,int line) const;
  void setGpioAction(SyGpioEvent::Type type,int line,const QString &filename);
  int gpioWatchdogTimeout(SyGpioEvent::Type type,int line) const;
  void setGpioWatchdogTimeout(SyGpioEvent::Type type,int line,int msec);
  QString gpioWatchdogAction(SyGpioEvent::Type type,int line) const;
  void setGpioWatchdogAction(SyGpioEvent::Type type,int line,
			     const QString &filename);
  bool gpioWatchdogState(SyGpioEvent::Type type,int line) const;
  void setGpioWatchdogState(SyGpioEvent::Type type,int line,bool state);
  QDateTime timestamp(SyGpioEvent::Type type,int line);
  void touchTimestamp(SyGpioEvent::Type type,int line);

 private:
  int entry_source_number;
  QString entry_log_dirs[2][SWITCHYARD_GPIO_BUNDLE_SIZE];
  QString entry_strings[2][ConfigEntry::LastRole][SWITCHYARD_GPIO_BUNDLE_SIZE];
  QString entry_actions[2][SWITCHYARD_GPIO_BUNDLE_SIZE];
  int entry_watchdog_timeouts[2][SWITCHYARD_GPIO_BUNDLE_SIZE];
  QString entry_watchdog_actions[2][SWITCHYARD_GPIO_BUNDLE_SIZE];
  bool entry_watchdog_states[2][SWITCHYARD_GPIO_BUNDLE_SIZE];
  QDateTime entry_timestamps[2][SWITCHYARD_GPIO_BUNDLE_SIZE];
};




class Config : public QObject
{
  Q_OBJECT;
 public:
  Config(QObject *parent=0);
  ConfigEntry *configEntry(unsigned srcnum) const;
  bool logActive(SyGpioEvent *e);
  QString logDir(SyGpioEvent *e);
  QString logFilename(SyGpioEvent *e);
  QString logLine(SyGpioEvent *e);
  bool scriptActive(SyGpioEvent *e);
  QString gpioAction(SyGpioEvent *e);
  void touchTimestamp(SyGpioEvent *e);
  bool load();
  static QString gpioTypeText(SyGpioEvent::Type type);
  static QString stateText(bool state);

 signals:
  void watchdogStateChanged(ConfigEntry *e,SyGpioEvent::Type type,int line,
			    bool state);

 private slots:
  void checkWatchdogsData();

 private:
  QString TypeString(SyGpioEvent::Type) const;
  QTimer *conf_watchdog_timer;
  ConfigEntry::Role GetRole(SyGpioEvent *e) const;
  std::map<unsigned,ConfigEntry *> conf_entries;
};


#endif  // LWLOGGER_H


src/lwlogger.h
// lwlogger.h
//
// lwlogger(8) LiveWire routing daemon
//
//   (C) Copyright 2016 Fred Gleason <fredg@paravelsystems.com>
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as
//   published by the Free Software Foundation; either version 2 of
//   the License, or (at your option) any later version.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public
//   License along with this program; if not, write to the Free Software
//   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

#ifndef LWLOGGER_H
#define LWLOGGER_H

#include <QList>
#include <QObject>
#include <QTimer>

#include <sy/sygpio_server.h>

#include "config.h"
#include "scriptevent.h"

#define LWLOGGER_USAGE "[options]\n"

class MainObject : public QObject
{
 Q_OBJECT;
 public:
  MainObject(QObject *parent=0);

 private slots:
  void gpioReceivedData(SyGpioEvent *e);
  void scriptFinishedData();
  void collectGarbageData();
  void watchdogStateChangedData(ConfigEntry *e,SyGpioEvent::Type type,
				int line,bool state);


 private:
  void RunGpioScript(const QString &pgm,const QString &logname,SyGpioEvent *e);
  void RunWatchdogScript(const QString &pgm,const QString &logname,
			 int srcnum,SyGpioEvent::Type type,int line,bool state);
  void Log(const QString &dirname,const QString &msg) const;
  QList<ScriptEvent *> main_script_list;
  QTimer *main_garbage_timer;
  SyGpioServer *main_gpio_server;
  Config *main_config;
};


#endif  // LWLOGGER_H


src/lwlogger.cpp
// lwlogger.cpp
//
// lwlogger(8) LiveWire routing daemon
//
//   (C) Copyright 2016 Fred Gleason <fredg@paravelsystems.com>
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as
//   published by the Free Software Foundation; either version 2 of
//   the License, or (at your option) any later version.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public
//   License along with this program; if not, write to the Free Software
//   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

#include <stdio.h>
#include <syslog.h>

#include <QCoreApplication>
#include <QDateTime>
#include <QStringList>

#include "lwlogger.h"

MainObject::MainObject(QObject *parent)
  : QObject(parent)
{
  main_config=new Config();
  main_config->load();
  connect(main_config,
	 SIGNAL(watchdogStateChanged(ConfigEntry *,SyGpioEvent::Type,int,bool)),
	  this,
     SLOT(watchdogStateChangedData(ConfigEntry *,SyGpioEvent::Type,int,bool)));

  main_gpio_server=new SyGpioServer(new SyRouting(0,0),this);
  connect(main_gpio_server,SIGNAL(gpioReceived(SyGpioEvent *)),
	  this,SLOT(gpioReceivedData(SyGpioEvent *)));

  main_garbage_timer=new QTimer(this);
  main_garbage_timer->setSingleShot(true);
  connect(main_garbage_timer,SIGNAL(timeout()),this,SLOT(collectGarbageData()));
}


void MainObject::gpioReceivedData(SyGpioEvent *e)
{
  main_config->touchTimestamp(e);
  if(main_config->logActive(e)) {
    Log(main_config->logDir(e),main_config->logLine(e));
  }
  if(main_config->scriptActive(e)) {
    RunGpioScript(main_config->gpioAction(e),main_config->logFilename(e),e);
  }
}


void MainObject::scriptFinishedData()
{
  main_garbage_timer->start(1);
}


void MainObject::collectGarbageData()
{
  for(int i=main_script_list.size()-1;i>=0;i--) {
    if(main_script_list.at(i)->processState()==QProcess::NotRunning) {
      delete main_script_list.at(i);
      main_script_list.erase(main_script_list.begin()+i);
    }
  }
}


void MainObject::watchdogStateChangedData(ConfigEntry *e,SyGpioEvent::Type type,
					  int line,bool state)
{
  ConfigEntry::Role role=ConfigEntry::WatchdogSet;
  if(!state) {
    role=ConfigEntry::WatchdogReset;
  }
  if(!e->gpioString(type,role,line).isEmpty()) {
    Log(e->logDir(type,line),e->gpioString(type,role,line));
  }
  if(!e->gpioWatchdogAction(type,line).isEmpty()) {
    RunWatchdogScript(e->gpioWatchdogAction(type,line),e->logDir(type,line),
		      e->sourceNumber(),type,line,state);
  }
}


void MainObject::RunGpioScript(const QString &pgm,const QString &logname,
			       SyGpioEvent *e)
{
  /*
  printf("Run script: %s\n",(const char *)pgm.toUtf8());
  printf("   Logname: %s\n",(const char *)logname.toUtf8());
  printf("      GPIO: %s\n",(const char *)e->dump().toUtf8());
  */
  QStringList args;
  args.push_back(Config::gpioTypeText(e->type()));
  args.push_back(QString().sprintf("%d",e->sourceNumber()));
  args.push_back(QString().sprintf("%d",e->line()+1));
  args.push_back(Config::stateText(e->state()));
  args.push_back(Config::stateText(e->isPulse()));
  args.push_back(e->originAddress().toString());
  args.push_back(QString().sprintf("%d",0xFFFF&e->originPort()));

  main_script_list.push_back(new ScriptEvent(this));
  connect(main_script_list.back(),SIGNAL(finished()),
	  this,SLOT(scriptFinishedData()));
  main_script_list.back()->start(pgm,args,logname);
}


void MainObject::RunWatchdogScript(const QString &pgm,const QString &logname,
				   int srcnum,SyGpioEvent::Type type,int line,
				   bool state)
{
  QStringList args;
  args.push_back(Config::gpioTypeText(type));
  args.push_back(QString().sprintf("%d",srcnum));
  args.push_back(QString().sprintf("%d",line+1));
  args.push_back(Config::stateText(state));

  main_script_list.push_back(new ScriptEvent(this));
  connect(main_script_list.back(),SIGNAL(finished()),
	  this,SLOT(scriptFinishedData()));
  main_script_list.back()->start(pgm,args,logname);
}


void MainObject::Log(const QString &dirname,const QString &msg) const
{
  FILE *f=NULL;
  QDateTime now=QDateTime::currentDateTime();
  int tenth=lround((double)now.time().msec()/100.0);
  if(tenth==10) {
    now.setTime(now.time().addSecs(1));
    tenth=0;
  }
  QString filename=dirname+"/"+now.toString("yyyy-MM-dd")+".txt";
  
  if((f=fopen(filename.toUtf8(),"a"))==NULL) {
    syslog(LOG_WARNING,"unable to open log file \"%s\"",
	   (const char *)filename.toUtf8());
    return;
  }
  fprintf(f,"%s: %s\n",(const char *)(now.toString("hh:mm:ss")+
				      QString().sprintf(".%d",tenth)).toUtf8(),
	  (const char *)msg.toUtf8());
  fclose(f);
}


int main(int argc,char *argv[])
{
  QCoreApplication a(argc,argv);
  new MainObject();

  return a.exec();
}


src/scriptevent.h
// scriptevent.h
//
// Execute a script
//
//   (C) Copyright 2017 Fred Gleason <fredg@paravelsystems.com>
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as
//   published by the Free Software Foundation; either version 2 of
//   the License, or (at your option) any later version.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public
//   License along with this program; if not, write to the Free Software
//   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

#ifndef SCRIPTEVENT_H
#define SCRIPTEVENT_H

#include <QObject>
#include <QProcess>

class ScriptEvent : public QObject
{
  Q_OBJECT;
 public:
  ScriptEvent(QObject *parent=0);
  ~ScriptEvent();
  QProcess::ProcessState processState() const;

 signals:
  void finished();

 public slots:
  void start(const QString &pgm,const QStringList &args,const QString &logfile);

 private slots:
  void processFinishedData(int exit_code,QProcess::ExitStatus status);
  void processErrorData(QProcess::ProcessError err);

 private:
  void Log(const QString &msg) const;
  QProcess *script_process;
  QString script_program;
  QString script_logfile;
};


#endif  // SCRIPTEVENT_H


src/scriptevent.cpp
// scriptevent.cpp
//
// Execute a script
//
//   (C) Copyright 2017 Fred Gleason <fredg@paravelsystems.com>
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as
//   published by the Free Software Foundation; either version 2 of
//   the License, or (at your option) any later version.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public
//   License along with this program; if not, write to the Free Software
//   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

#include <stdio.h>

#include <QDateTime>

#include "scriptevent.h"

ScriptEvent::ScriptEvent(QObject *parent)
  : QObject(parent)
{
  script_process=new QProcess(this);
  connect(script_process,SIGNAL(finished(int,QProcess::ExitStatus)),
	  this,SLOT(processFinishedData(int,QProcess::ExitStatus)));
  connect(script_process,SIGNAL(error(QProcess::ProcessError)),
	  this,SLOT(processErrorData(QProcess::ProcessError)));
}


ScriptEvent::~ScriptEvent()
{
  delete script_process;
}


QProcess::ProcessState ScriptEvent::processState() const
{
  return script_process->state();
}


void ScriptEvent::start(const QString &pgm,const QStringList &args,
			const QString &logfile)
{
  script_program=pgm;
  script_logfile=logfile;
  script_process->start(pgm,args);
}


void ScriptEvent::processFinishedData(int exit_code,QProcess::ExitStatus status)
{
  Log(script_process->readAllStandardOutput());

  if(status!=QProcess::NormalExit) {
    Log("script \""+script_program+"\" crashed");
  }
  else {
    if(exit_code!=0) {
      Log("script \""+script_program+"\" returned exit code "+
	  QString().sprintf("%d",exit_code));
    }
  }
  emit finished();
}


void ScriptEvent::processErrorData(QProcess::ProcessError err)
{
  Log("script \""+script_program+"\" generated process error "+
      QString().sprintf("%d",err));
}


void ScriptEvent::Log(const QString &msg) const
{
  if(!msg.trimmed().isEmpty()) {
    FILE *f=NULL;

    if((f=fopen(script_logfile.toUtf8(),"a"))!=NULL) {
      fprintf(f,"%s: ",
	      (const char *)QTime::currentTime().toString("hh:mm:ss").toUtf8());
      fprintf(f,"%s",(const char *)msg.toUtf8());
      if(msg.right(1)!="\n") {
	fprintf(f,"\n");
      }
      fclose(f);
    }
  }
}


src/config.cpp
// config.cpp
//
// Configuration for lwlogger(8)
//
//   (C) Copyright 2016 Fred Gleason <fredg@paravelsystems.com>
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as
//   published by the Free Software Foundation; either version 2 of
//   the License, or (at your option) any later version.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public
//   License along with this program; if not, write to the Free Software
//   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

#include <sy/syprofile.h>

#include "config.h"

ConfigEntry::ConfigEntry(int srcnum)
{
  entry_source_number=srcnum;
  QDateTime now(QDateTime::currentDateTime());
  for(int i=0;i<2;i++) {
    for(int j=0;j<SWITCHYARD_GPIO_BUNDLE_SIZE;j++) {
      entry_watchdog_states[i][j]=false;
      entry_timestamps[i][j]=now;
    }
  }
}


ConfigEntry *Config::configEntry(unsigned srcnum) const
{
  try {
    return conf_entries.at(srcnum);
  }
  catch(...)
    {
    }
  return NULL;
}


int ConfigEntry::sourceNumber() const
{
  return entry_source_number;
}


QString ConfigEntry::logDir(SyGpioEvent::Type type,int line) const
{
  return entry_log_dirs[type][line];
}


void ConfigEntry::setLogDir(SyGpioEvent::Type type,int line,
			    const QString &str)
{
  entry_log_dirs[type][line]=str;
}


QString ConfigEntry::gpioString(SyGpioEvent::Type type,ConfigEntry::Role role,
				int line)
{
  return entry_strings[type][role][line];
}


    void ConfigEntry::setGpioString(SyGpioEvent::Type type,
				    ConfigEntry::Role role,int line,
				    const QString &str)
{
  entry_strings[type][role][line]=str;
}


QString ConfigEntry::gpioAction(SyGpioEvent::Type type,int line) const
{
  return entry_actions[type][line];
}


void ConfigEntry::setGpioAction(SyGpioEvent::Type type,int line,
				const QString &filename)
{
  entry_actions[type][line]=filename;
}


int ConfigEntry::gpioWatchdogTimeout(SyGpioEvent::Type type,int line) const
{
  return entry_watchdog_timeouts[type][line];
}


void ConfigEntry::setGpioWatchdogTimeout(SyGpioEvent::Type type,int line,
					 int msec)
{
  entry_watchdog_timeouts[type][line]=msec;
}


QString ConfigEntry::gpioWatchdogAction(SyGpioEvent::Type type,int line) const
{
  return entry_watchdog_actions[type][line];
}


void ConfigEntry::setGpioWatchdogAction(SyGpioEvent::Type type,int line,
					const QString &filename)
{
  entry_watchdog_actions[type][line]=filename;
}


bool ConfigEntry::gpioWatchdogState(SyGpioEvent::Type type,int line) const
{
  return entry_watchdog_states[type][line];
}


void ConfigEntry::setGpioWatchdogState(SyGpioEvent::Type type,int line,
				       bool state)
{
  entry_watchdog_states[type][line]=state;
}


QDateTime ConfigEntry::timestamp(SyGpioEvent::Type type,int line)
{
  return entry_timestamps[type][line];
}


void ConfigEntry::touchTimestamp(SyGpioEvent::Type type,int line)
{
  entry_timestamps[type][line].setDate(QDate::currentDate());
  entry_timestamps[type][line].setTime(QTime::currentTime());
}




Config::Config(QObject *parent)
  : QObject(parent)
{
  conf_watchdog_timer=new QTimer(this);
  connect(conf_watchdog_timer,SIGNAL(timeout()),
	  this,SLOT(checkWatchdogsData()));
}


bool Config::logActive(SyGpioEvent *e)
{
  ConfigEntry *entry=NULL;

  if((entry=configEntry(e->sourceNumber()))==NULL) {
    return false;
  }
  return (!entry->logDir(e->type(),e->line()).isEmpty())&&
    (!entry->gpioString(e->type(),GetRole(e),e->line()).isEmpty());
}


QString Config::logDir(SyGpioEvent *e)
{
  return conf_entries.at(e->sourceNumber())->logDir(e->type(),e->line());
}


QString Config::logFilename(SyGpioEvent *e)
{
  return logDir(e)+"/"+QDate::currentDate().toString("yyyy-MM-dd")+".txt";
}


bool Config::scriptActive(SyGpioEvent *e)
{
  ConfigEntry *entry=NULL;

  if((entry=configEntry(e->sourceNumber()))==NULL) {
    return false;
  }
  return !entry->gpioAction(e->type(),e->line()).isEmpty();
}


QString Config::gpioAction(SyGpioEvent *e)
{
  return conf_entries.at(e->sourceNumber())->gpioAction(e->type(),e->line());
}


void Config::touchTimestamp(SyGpioEvent *e)
{
  try {
    conf_entries.at(e->sourceNumber())->touchTimestamp(e->type(),e->line());
  }
  catch(...)
    {
    }
}


QString Config::logLine(SyGpioEvent *e)
{
  return conf_entries.at(e->sourceNumber())->
    gpioString(e->type(),GetRole(e),e->line());
}


bool Config::load()
{
  SyProfile *p=new SyProfile();
  p->setSource(CONFIG_FILE);
  QStringList sections=p->sectionNames();
  bool ok=false;

  for(int i=0;i<sections.size();i++) {
    unsigned srcnum=sections.at(i).toUInt(&ok);
    if(ok&&(srcnum>0)&&(srcnum<32768)) {
      conf_entries[srcnum]=new ConfigEntry(srcnum);
      ConfigEntry *e=conf_entries.at(srcnum);
      for(int j=0;j<2;j++) {
	SyGpioEvent::Type type=(SyGpioEvent::Type)j;
	for(int k=0;k<SWITCHYARD_GPIO_BUNDLE_SIZE;k++) {
	  e->setLogDir(type,k,p->stringValue(sections.at(i),TypeString(type)+
					    QString().sprintf("%dLogDir",k+1)));
	  e->setGpioString(type,ConfigEntry::Off,k,
			   p->stringValue(sections.at(i),TypeString(type)+
					  QString().sprintf("%dOff",k+1)));
	  e->setGpioString(type,ConfigEntry::On,k,
			   p->stringValue(sections.at(i),TypeString(type)+
					  QString().sprintf("%dOn",k+1)));
	  e->setGpioString(type,ConfigEntry::Pulse,k,
			   p->stringValue(sections.at(i),TypeString(type)+
					  QString().sprintf("%dPulse",k+1)));
	  e->setGpioString(type,ConfigEntry::WatchdogSet,k,
			   p->stringValue(sections.at(i),TypeString(type)+
				     QString().sprintf("%dWatchdogSet",k+1)));
	  e->setGpioString(type,ConfigEntry::WatchdogReset,k,
			   p->stringValue(sections.at(i),TypeString(type)+
				     QString().sprintf("%dWatchdogReset",k+1)));
	  e->setGpioAction(type,k,
			   p->stringValue(sections.at(i),TypeString(type)+
					  QString().sprintf("%dAction",k+1)));
	  e->setGpioWatchdogTimeout(type,k,
		   p->intValue(sections.at(i),TypeString(type)+
			       QString().sprintf("%dWatchdogTimeout",k+1)));
	  e->setGpioWatchdogAction(type,k,
       		   p->stringValue(sections.at(i),TypeString(type)+
				  QString().sprintf("%dWatchdogAction",k+1)));
	}
      }
    }
  }
  conf_watchdog_timer->start(1000);
  return true;
}


QString Config::gpioTypeText(SyGpioEvent::Type type)
{
  if(type==SyGpioEvent::TypeGpi) {
    return QString("gpi");
  }
  return QString("gpo");
}


QString Config::stateText(bool state)
{
  if(state) {
    return QString("true");
  }
  return QString("false");
}


void Config::checkWatchdogsData()
{
  QDateTime now=QDateTime::currentDateTime();

  for(std::map<unsigned,ConfigEntry *>::const_iterator it=conf_entries.begin();
      it!=conf_entries.end();it++) {
    ConfigEntry *ce=it->second;
    for(int i=0;i<2;i++) {
      SyGpioEvent::Type type=(SyGpioEvent::Type)i;
      for(int j=0;j<SWITCHYARD_GPIO_BUNDLE_SIZE;j++) {
	if(ce->gpioWatchdogTimeout(type,j)>0) {
	  if((!ce->gpioWatchdogState(type,j))&&
	     (ce->timestamp(type,j).addSecs(ce->gpioWatchdogTimeout(type,j))<
	      now)) {
	    ce->setGpioWatchdogState(type,j,true);
	    emit watchdogStateChanged(ce,type,j,true);
	  }
	  if(ce->gpioWatchdogState(type,j)&&
	     (ce->timestamp(type,j).addSecs(ce->gpioWatchdogTimeout(type,j))>
	      now)) {
	    ce->setGpioWatchdogState(type,j,false);
	    emit watchdogStateChanged(ce,type,j,false);
	  }
	}
      }
    }
  }
}


ConfigEntry::Role Config::GetRole(SyGpioEvent *e) const
{
  if(!e->isPulse()) {
    if(e->state()) {
      return ConfigEntry::On;
    }
    return ConfigEntry::Off;
  }
  return ConfigEntry::Pulse;
}


QString Config::TypeString(SyGpioEvent::Type type) const
{
  if(type==SyGpioEvent::TypeGpi) {
    return "Gpi";
  }
  return "Gpo";
}
